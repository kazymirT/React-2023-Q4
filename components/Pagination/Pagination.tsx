import { PaginationPropsType } from '../type/type';

export const Pagination = (props: PaginationPropsType) => {
  const totalPages: number = Math.ceil(
    Number(props.total) / Number(props.limit)
  );
  const currentPage: number = props.page === 0 || isNaN(props.page) ? 1 : props.page;
  const handlePrevPage = () => {
    props.onChange(currentPage - 1, totalPages);
  };
  const handleNextPage = () => {
    props.onChange(currentPage + 1, totalPages);
  };

  return (
    <div className="pagination">
      <button
        onClick={handlePrevPage}
        data-testid="btn-prev"
        disabled={currentPage === 1}
      >
        &#8656;
      </button>
      <span>
        {currentPage}/{totalPages}
      </span>
      <button
        onClick={handleNextPage}
        data-testid="btn-next"
        disabled={currentPage === totalPages}
      >
        &#8658;
      </button>
    </div>
  );
};
