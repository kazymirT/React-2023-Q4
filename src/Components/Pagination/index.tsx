import { PaginationProps } from '../../type/type';

export const Pagination = (props: PaginationProps) => {
  const handlePageChange = (newPage: number) => {
    if (newPage >= 1 && newPage <= props.totalPages) {
      props.onPageChange(newPage);
    }
  };

  return (
    <div className="pagination">
      <button
        onClick={() => handlePageChange(props.currentPage - 1)}
        disabled={props.currentPage === 1}
      >
        Назад
      </button>
      <span>
        Сторінка {props.currentPage} з {props.totalPages}
      </span>
      <button
        onClick={() => handlePageChange(props.currentPage + 1)}
        disabled={props.currentPage === props.totalPages}
      >
        Вперед
      </button>
    </div>
  );
};
