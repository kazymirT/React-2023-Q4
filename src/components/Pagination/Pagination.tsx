import React from "react";
import { PaginationPropsType } from "../type/type";

export const Pagination: React.FC<PaginationPropsType> = ({
  limit,
  page,
  total,
  onChange,
}) => {
  const skip: number = (Number(page) - 1) * Number(limit);
  const currentPage: number = Number(skip) / Number(limit) + 1;
  const totalPages: number = Math.ceil(Number(total) / Number(limit));

  const handlePrevPage = () => {
    onChange(currentPage - 1, totalPages);
  };
  const handleNextPage = () => {
    onChange(currentPage + 1, totalPages);
  };

  return (
    <div className="pagination" data-testid={"pagination"}>
      <button
        onClick={handlePrevPage}
        data-testid="btn-prev"
        disabled={currentPage === 1}
      >
        &#8656;
      </button>
      <span data-testid={"pagination-page"}>
        {currentPage}/{totalPages}
      </span>
      <button
        onClick={handleNextPage}
        data-testid="btn-next"
        disabled={currentPage === totalPages}
      >
        &#8658;
      </button>
    </div>
  );
};
