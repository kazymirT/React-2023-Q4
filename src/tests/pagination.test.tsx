import React from "react";
import { describe, expect, it, vi } from "vitest";
import "@testing-library/jest-dom";
import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import mockRouter from "next-router-mock";
import { MemoryRouterProvider } from "next-router-mock/MemoryRouterProvider";
import { Pagination } from "@/components/Pagination/Pagination";

vi.mock("next/router", () => vi.importActual("next-router-mock"));

describe("Tests for the Pagination component", () => {
  it("The component updates URL query parameter when page changes", async () => {
    const limit = 10;
    const total = 30;
    const page = 2;
    const skip: number = (page - 1) * limit;
    const currentPage = Number(skip) / Number(limit) + 1;
    const totalPages = Math.ceil(total / limit);
    const onChange = vi.fn();

    mockRouter.push("/");
    render(
      <Pagination
        limit={limit}
        page={page}
        total={total}
        onChange={onChange}
      />,
      {
        wrapper: MemoryRouterProvider,
      },
    );
    expect(screen.getByTestId("pagination")).toMatchSnapshot();
    expect(await screen.findByTestId("pagination")).toBeInTheDocument();
    expect(await screen.findByTestId("btn-prev")).toBeInTheDocument();
    expect(await screen.findByTestId("btn-next")).toBeInTheDocument();
    expect(await screen.findByTestId("pagination-page")).toHaveTextContent(
      `${currentPage}/${totalPages}`,
    );

    fireEvent.click(screen.getByTestId("btn-next"));
    expect(onChange).toHaveBeenCalledWith(Number(page) + 1, totalPages);

    fireEvent.click(screen.getByTestId("btn-prev"));
    expect(onChange).toHaveBeenCalledWith(Number(page) - 1, totalPages);
    cleanup();
  });
});
