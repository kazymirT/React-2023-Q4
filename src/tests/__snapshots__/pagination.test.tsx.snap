// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for the Pagination component > The component updates URL query parameter when page changes 1`] = `
<div
  class="pagination"
  data-testid="pagination"
>
  <button
    data-testid="btn-prev"
  >
    ⇐
  </button>
  <span
    data-testid="pagination-page"
  >
    2
    /
    3
  </span>
  <button
    data-testid="btn-next"
  >
    ⇒
  </button>
</div>
`;
